### ISSUE TITLE
XRay Remote Service Name is always "Remote"

### ISSUE THREAD

**Issue by adriancole** [openzipkin/zipkin-aws#58]
_From @cemo on October 18, 2017 21:20_

I am in the process of customizing brave and putting into the production system but I came across a problem.

Despite of creating my interceptor with hardcoded labels, it always displays *remote* in the console. I could not find time to check it but seems there is a bug at there. I might give a try tomorrow to find culprit. 

![image](https://user-images.githubusercontent.com/44156/31743108-946c32ae-b462-11e7-8476-59bbd5938b6d.png)

_Copied from original issue: openzipkin/brave#524_

**Comment by cemo** [comment]
@adriancole seems this part should be splitted into if/else.

https://github.com/openzipkin/zipkin-aws/blob/master/storage-xray-udp/src/main/java/zipkin2/storage/xray_udp/UDPMessageEncoder.java#L57
```
      if (span.kind() != null) writer.name("namespace").value("remote");
```

to
```
      if (span.kind() != null) {
           if(span.remoteEndpoint() != null && span.remoteEndpoint().serviceName() != null ){
                 writer.name("namespace").value(span.remoteEndpoint().serviceName());
           }
           else {
                writer.name("namespace").value("remote");
           }
      }

```

Would you accept a PR from me? :)

**Comment by adriancole** [comment]
Lemme just check semantics of namespace quickly

**Comment by adriancole** [comment]
Looks like the name should be set in the condition where we check span.kind

If span.kind != null
  set subsegment
  set namespace=remote
  set name = remoteServiceName
Else
  set name = localServiceName


http://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html#api-segmentdocuments-subsegments

On 19 Oct 2017 22:23, "Cemalettin Koc" <notifications@github.com> wrote:

@adriancole <https://github.com/adriancole> seems this part should be
splitted into if/else.

https://github.com/openzipkin/zipkin-aws/blob/master/
storage-xray-udp/src/main/java/zipkin2/storage/xray_udp/
UDPMessageEncoder.java#L57

      if (span.kind() != null) writer.name("namespace").value("remote");

to

      if (span.kind() != null) {
           if(span.remoteEndpoint() != null &&
span.remoteEndpoint().serviceName() != null ){

writer.name("namespace").value(span.remoteEndpoint().serviceName());
           }
           else {
                writer.name("namespace").value("remote");
           }
      }


Would you accept a PR from me? :)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openzipkin/zipkin-aws/issues/58#issuecomment-338010949>,
or mute the thread

**Comment by cemo** [comment]
Seems you are right. I will give a try as early as possible.

**Comment by cemo** [comment]
@adriancole sorry for late reply. I have checked and [modified] as you wrote. Seems "remote" is mandatory. It starts to become visible when I set "http.url". I have checked brave's code but I could not see a place where "http.url" is set. Here is the result:

![image](https://user-images.githubusercontent.com/44156/32398482-43f12d64-c0f8-11e7-8660-3ee6b9a25532.png)


One another missing part is `user` of segment. How should I set it? 
> user – A string that identifies the user who sent the request.

seems beneficial. I could not see user in Encoder. 
 
[modified]: https://gist.github.com/cemo/40efc6aec0d3e3f6220fbc4eb506f63f

**Comment by adriancole** [comment]
No worries.. seems like we might need a default http parser for x-ray, to
pick the right fields. Ex http.url is not default

https://github.com/openzipkin/brave/blob/master/instrumentation/http/README.md#span-data-policy

Wanna try to make one for client and server that picks the best things?

**Comment by cemo** [comment]
I would be happy if I can contribute. Do you mean to add a new HttpServerParser and HttpClientParser `xray` specific?

**Comment by cemo** [comment]
@adriancole Should I add a new module as `zipkin-aws-http`?

**Comment by adriancole** [comment]
>
> I would be happy if I can contribute. Do you mean to add a new
> HttpServerParser and HttpClientParser xray specific?
>
yeah, it will probably end up as brave-instrumentation-xray or something
(because these types are brave specific)

**Comment by cemo** [comment]
@adriancole I had some difficulties to grab responsibilities of the project and modules. There is `zipkin', `zipkin-reporter-java`, `brave`, `zipkin-aws`.  Since I had some difficulties in modules I could not be sure that this new module where it should be belongs to. In brave there is also [aws-propagation] module as well. It is dependency free and not including anything from `zipkin-aws`. Now we are evaluating adding a new module in this project as ` brave-instrumentation-xray` and it will has dependencies about brave. This can be easily chicken-egg problem since both module in future can be dependent to each other. I am suggesting to either to add this new module in brave or move [aws-propagation] to here which sound more logical at first glance to me. Please note that I am very new to this land and my limited knowledge may be not correct. 

[aws-propagation]: `https://github.com/openzipkin/brave/tree/master/propagation/aws`

**Comment by adriancole** [comment]
This can be easily chicken-egg problem since both module in future can be
> dependent to each other. I am suggesting to either to add this new module
> in brave or move aws-propagation to here which sound more logical at first
> glance to me. Please note that I am very new to this land and my limited
> knowledge may be not correct.
>
It was a bad suggestion of mine to have the brave code here, as this repo
is only about out-of-band architecture (transport (reporter/collector) and
storage). The reporter piece is here because zipkin-reporter isn't just
brave (other libraries use it).

Instrumentation (propagation, tagging policy etc) indeed live where the
code does, in brave. So, at the point we're ready to merge some helpers, it
would be in the brave repo, you're right. I wouldn't move propagation code
here, as it depends on brave classes.

Anyway I think you've got the relationships right, and indeed the UDP
component being here will feel a bit uncomfortable until things stabilize,
but on the other hand it is most like existing code here.

**Comment by cemo** [comment]
@adriancole I am in the process of supporting all remaining parts of x-ray and will change necessary places in Brave. I also made some progress on Brave as well. I will provide feature parity with XRay.

**Comment by adriancole** [comment]
thx for the trail blazing!

**Comment by cemo** [comment]
@adriancole I had some progress in Encoder part. Basically I added

1. Support for SQL
2. Error Handling 
3. Using Error Responses to visualize in UI. (You can see in attached screens below)
4. Option to override remote namespace
5. AWS internal components such as DynamoDB, SQS to visualize correctly

There might some parts as well I touched but now I will switch to Brave for refactoring parts you suggested. You can review changes at https://github.com/openzipkin/zipkin-aws/pull/59

Some parts are still requiring to be polished especially `cause` related parts.

![image](https://user-images.githubusercontent.com/44156/32536389-fe5f75ea-c466-11e7-952b-bb8db38ddaf1.png)

![image](https://user-images.githubusercontent.com/44156/32536403-0cde12fc-c467-11e7-87c4-afddc83868f1.png)


I think that it is enough for a start.

**Comment by adriancole** [comment]
@cemo champion!

**Comment by trajano** [comment]
I am thinking this may be related to my [SO question](https://stackoverflow.com/questions/71088991/how-do-i-map-http-path-in-zipkin-to-aws-xray-request-url?noredirect=1&lq=1)
